File: typescript/tests/commerce_payfast_board.test.ts
// commerce_payfast_board.test.ts
// Verifies the CircleAI.Commerce.Integration.PayFast port. The signature MD5s
// and the WebUtility.UrlEncode output are pinned to values captured from the
// real .NET 10 runtime (see src/commerce/integration/payfast/index.ts header).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryPayFastBoard,
  CommerceIntegrationPayFastDomainContext,
  webUtilityUrlEncode,
  payFastConfig,
  payFastItnPayload,
} from "../src/commerce/integration/payfast/index";

describe("webUtilityUrlEncode (.NET parity)", () => {
  it("matches .NET WebUtility.UrlEncode byte-for-byte on the probe string", () => {
    const input = "AB az 09 -_.!*()'~ +&=%@#$/:,;? \"<>{}| é";
    const expected = "AB+az+09+-_.!*()%27%7E+%2B%26%3D%25%40%23%24%2F%3A%2C%3B%3F+%22%3C%3E%7B%7D%7C+%C3%A9";
    assert.equal(webUtilityUrlEncode(input), expected);
  });

  it("space becomes '+', unreserved -_.!*() pass through, ' and ~ are encoded", () => {
    assert.equal(webUtilityUrlEncode("a b"), "a+b");
    assert.equal(webUtilityUrlEncode("-_.!*()"), "-_.!*()");
    assert.equal(webUtilityUrlEncode("'"), "%27");
    assert.equal(webUtilityUrlEncode("~"), "%7E");
  });
});

describe("InMemoryPayFastBoard.signatureFor (.NET parity)", () => {
  function board(passphrase: string) {
    return new InMemoryPayFastBoard(payFastConfig("10000100", "key", passphrase, true));
  }

  it("no passphrase — pinned MD5", () => {
    const fields = new Map<string, string>([
      ["merchant_id", "10000100"],
      ["item_name", "Test Product"],
      ["amount", "100.00"],
    ]);
    assert.equal(board("").signatureFor(fields), "a928072455fff91c7bb4238393006983");
  });

  it("with passphrase — appended and hashed, pinned MD5", () => {
    const fields = new Map<string, string>([
      ["merchant_id", "10000100"],
      ["item_name", "Test Product"],
      ["amount", "100.00"],
    ]);
    assert.equal(board("my secret pass").signatureFor(fields), "974e3c4c1f4d3d21a8e357f799e2896c");
  });

  it("values containing & and = are encoded before hashing — pinned MD5", () => {
    const fields = new Map<string, string>([
      ["a", "x&y"],
      ["b", "c=d"],
      ["e", "space here"],
    ]);
    assert.equal(board("").signatureFor(fields), "0b8fc27bd19fa06043bbf42d30da18eb");
  });

  it("single empty value, no passphrase — pinned MD5 (trailing & trimmed)", () => {
    const fields = new Map<string, string>([["k", ""]]);
    assert.equal(board("").signatureFor(fields), "7de211558c5b2e4d2d6d255f028a1e1a");
  });

  it("null orderedFields throws", () => {
    assert.throws(() => board("").signatureFor(null as never), /orderedFields required/);
  });
});

describe("InMemoryPayFastBoard — ITN + webhooks", () => {
  it("verifyItn matches on merchant id", () => {
    const b = new InMemoryPayFastBoard(payFastConfig("MID", "key", "", false));
    assert.equal(b.verifyItn(payFastItnPayload("MID", "pid", "COMPLETE", 100, "m1", "sig")), true);
    assert.equal(b.verifyItn(payFastItnPayload("OTHER", "pid", "COMPLETE", 100, "m1", "sig")), false);
  });

  it("recentWebhooks returns most-recent-first, limited", () => {
    const b = new InMemoryPayFastBoard(payFastConfig("MID", "key", "", false));
    for (let i = 1; i <= 5; i++) {
      b.recordWebhook(payFastItnPayload("MID", `p${i}`, "COMPLETE", i, `m${i}`, "sig"));
    }
    assert.deepEqual(
      b.recentWebhooks(3).map((w) => w.paymentId),
      ["p5", "p4", "p3"],
    );
    assert.equal(b.recentWebhooks().length, 5);
  });

  it("constructing with a null config throws", () => {
    assert.throws(() => new InMemoryPayFastBoard(null as never), /cfg required/);
  });

  it("domain context exposes prompt + compliance + tools", () => {
    assert.ok(CommerceIntegrationPayFastDomainContext.systemPromptSnippet.includes("[DOMAIN: Commerce.Integration.PayFast]"));
    assert.deepEqual(CommerceIntegrationPayFastDomainContext.complianceFlags, [
      "PCI_DSS",
      "POPIA",
      "PASA",
      "Consumer_Protection_Act",
    ]);
    assert.deepEqual(CommerceIntegrationPayFastDomainContext.suggestedTools, [
      "payfast_api",
      "webhook_debugger",
      "document_editor",
    ]);
  });
});

File: go/tests/commerce_payfast_board_test.go
// commerce_payfast_board_test.go
//
// Verifies the CircleAI.Commerce.Integration.PayFast port
// (commerce_payfast_board.go). The expected MD5 signatures are computed from the
// exact C# SignatureFor logic (WebUtility.UrlEncode(value).Replace("%20","+") +
// MD5 + lower-hex) run against the .NET 10 runtime, so this asserts byte-for-byte
// signature parity, plus ITN verify and reverse-chronological webhook recall.

package circleai_test

import (
	"testing"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestPayFast_SignatureParityWithPassphrase(t *testing.T) {
	board := circleai.NewInMemoryPayFastBoard(circleai.PayFastConfig{
		MerchantId: "10000100", MerchantKey: "46f0cd694581a", Passphrase: "myPassPhrase123", Sandbox: true,
	})
	fields := []circleai.PayFastField{
		{Key: "merchant_id", Value: "10000100"},
		{Key: "merchant_key", Value: "46f0cd694581a"},
		{Key: "return_url", Value: "https://example.com/return"},
		{Key: "item_name", Value: "Test Item & Co"},
		{Key: "amount", Value: "100.00"},
	}
	got := board.SignatureFor(fields)
	const want = "6bc713fc147ec994075f544b45b81999" // from C# reference
	if got != want {
		t.Fatalf("signature = %s, want %s", got, want)
	}
}

func TestPayFast_SignatureParityNoPassphrase(t *testing.T) {
	board := circleai.NewInMemoryPayFastBoard(circleai.PayFastConfig{MerchantId: "10000100"})
	fields := []circleai.PayFastField{
		{Key: "merchant_id", Value: "10000100"},
		{Key: "amount", Value: "55.50"},
		{Key: "item_name", Value: "Widget"},
	}
	if got, want := board.SignatureFor(fields), "46e389f8c122ba2946049c402fa54bb1"; got != want {
		t.Fatalf("no-pass signature = %s, want %s", got, want)
	}
}

func TestPayFast_SignatureSingleFieldTrailingAmpTrim(t *testing.T) {
	board := circleai.NewInMemoryPayFastBoard(circleai.PayFastConfig{MerchantId: "10000100"})
	fields := []circleai.PayFastField{{Key: "merchant_id", Value: "10000100"}}
	if got, want := board.SignatureFor(fields), "036c31e640eea59940d54b803a3473c6"; got != want {
		t.Fatalf("single-field signature = %s, want %s", got, want)
	}
}

func TestPayFast_SignatureTildeEncoding(t *testing.T) {
	// '~' must encode as %7E (WebUtility.UrlEncode), not stay literal (RFC 3986).
	board := circleai.NewInMemoryPayFastBoard(circleai.PayFastConfig{Passphrase: "pp~1"})
	fields := []circleai.PayFastField{{Key: "note", Value: "a~b c"}}
	if got, want := board.SignatureFor(fields), "1a0678443c390a08a9ab652127f409c2"; got != want {
		t.Fatalf("tilde signature = %s, want %s", got, want)
	}
}

func TestPayFast_VerifyItnAndConfig(t *testing.T) {
	cfg := circleai.PayFastConfig{MerchantId: "10000100", MerchantKey: "k", Passphrase: "p", Sandbox: false}
	board := circleai.NewInMemoryPayFastBoard(cfg)
	if board.Config().MerchantId != "10000100" || board.Config().Sandbox {
		t.Fatalf("config not returned: %+v", board.Config())
	}
	if !board.VerifyItn(circleai.PayFastItnPayload{MerchantId: "10000100"}) {
		t.Fatalf("matching merchant should verify")
	}
	if board.VerifyItn(circleai.PayFastItnPayload{MerchantId: "99999999"}) {
		t.Fatalf("mismatched merchant must not verify")
	}
}

func TestPayFast_RecentWebhooksReverseChronological(t *testing.T) {
	board := circleai.NewInMemoryPayFastBoard(circleai.PayFastConfig{MerchantId: "m"})
	for _, id := range []string{"w1", "w2", "w3"} {
		board.RecordWebhook(circleai.PayFastItnPayload{MerchantId: "m", PaymentId: id})
	}
	recent := board.RecentWebhooks(20)
	if len(recent) != 3 || recent[0].PaymentId != "w3" || recent[1].PaymentId != "w2" || recent[2].PaymentId != "w1" {
		t.Fatalf("recent webhooks reverse order failed: %+v", recent)
	}
	capped := board.RecentWebhooks(2)
	if len(capped) != 2 || capped[0].PaymentId != "w3" || capped[1].PaymentId != "w2" {
		t.Fatalf("capped recent failed: %+v", capped)
	}
}
